drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,543 @@
+	}
+	return 0;
+}
+
+static int eeprom_emu_range_is_valid(const struct device *dev, off_t address,
+				     size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	if ((address + len) <= dev_config->size) {
+		return 1;
+	}
+
+	return 0;
+}
+
+/*
+ * eeprom_emu_rw: read and write routine
+ */
+static int eeprom_emu_rw(const struct device *dev, off_t address, void *data,
+			 size_t len, bool write_op)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	uint8_t *data8 = (uint8_t *)data;
+	uint8_t buf[dev_config->flash_cbs];
+	off_t addr_jmp;
+
+	if ((!dev) || (!data) ||
+	    (!eeprom_emu_range_is_valid(dev, address, len))) {
+		return -EINVAL;
+	}
+
+	if ((write_op) && (dev_config->readonly)) {
+		LOG_WRN("attempt to write to read-only device");
+		return -EACCES;
+	}
+
+	LOG_DBG("EEPROM %s at [0x%tx] length[%d]", write_op ? "write" : "read",
+		(ptrdiff_t)address, len);
+
+	k_mutex_lock(&dev_data->lock, K_FOREVER);
+
+	addr_jmp = address & (sizeof(buf) - 1);
+	if (addr_jmp) {
+		/* unaligned start */
+		size_t plen = sizeof(buf) - addr_jmp;
+		(void)eeprom_emu_word_read(dev, address - addr_jmp, buf);
+		if (len < plen) {
+			plen = len;
+		}
+		if (write_op) {
+			memcpy(buf + addr_jmp, data8, plen);
+			eeprom_emu_word_write(dev, address - addr_jmp, buf);
+		} else {
+			memcpy(data8, buf + addr_jmp, plen);
+		}
+		len -= plen;
+		data8 += plen;
+		address += plen;
+	}
+	while (len >= sizeof(buf)) {
+		/* aligned part */
+		if (write_op) {
+			eeprom_emu_word_write(dev, address, data8);
+		} else {
+			eeprom_emu_word_read(dev, address, buf);
+			memcpy(data8, buf, sizeof(buf));
+		}
+		data8 += sizeof(buf);
+		address += sizeof(buf);
+		len -= sizeof(buf);
+	}
+	if (len) {
+		/* unaligned end */
+		(void)eeprom_emu_word_read(dev, address, buf);
+		if (write_op) {
+			memcpy(buf, data8, len);
+			eeprom_emu_word_write(dev, address, buf);
+		} else {
+			memcpy(data8, buf, len);
+		}
+	}
+
+	k_mutex_unlock(&dev_data->lock);
+
+	return 0;
+}
+
+static int eeprom_emu_read(const struct device *dev, off_t address, void *data,
+			   size_t len)
+{
+	return eeprom_emu_rw(dev, address, data, len, false);
+}
+
+static int eeprom_emu_write(const struct device *dev, off_t address,
+			    const void *data, size_t len)
+{
+	return eeprom_emu_rw(dev, address, (void *)data, len, true);
+}
+
+static size_t eeprom_emu_size(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	if (!dev) {
+		return 0;
+	}
+	return dev_config->size;
+}
+
+static int eeprom_emu_init(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	off_t offset;
+	uint8_t buf[dev_config->flash_cbs];
+
+	k_mutex_init(&dev_data->lock);
+	if (!device_is_ready(dev_config->flash_dev)) {
+		__ASSERT(0, "Could not get flash device binding");
+	}
+
```

Comment:
Since this is a software-only implementation of an EEPROM-like flash filesystem, I think we should add a version field to the stored data. Otherwise, we will never be able to change the flash representation.

Reply:
At first I was thinking, yes I can add this to the last record that gets overwritten with all 0. But then, the configuration is all done by the dts, there is no software configuration of the flash representation. Making changes to the dts is like making hardware changes, no compatibility guaranteed. That said, would it be possible to convert from one eeprom_emulation to a different one: yes, create two devices (different version) using the same partition: read in to a ram buffer using the old representation, erase the partition and write the ram buffer using the new representation. This method could even be used when making changes to the dts file.